### ISSUE TITLE
NUnitLite calls .NET Core 2.0 “.NET Core 4.0.0.0”

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2435]
> NUnitLite 3.9.0 (.NET Standard 1.6)
> Copyright (c) 2017 Charlie Poole, Rob Prouse
> Runtime Environment
>    OS Version: Microsoft Windows 6.3.9600 
>   CLR Version: .NET Core 4.0.0.0

Been bugging me.

**Comment by CharliePoole** [comment]
Most likely needs code in RuntimeFramework. It has a pot of built-in knowledge about what to call things and was probably never updated for .NET Core.

**Comment by mikkelbu** [comment]
Currently, we just ask `RuntimeInformation.FrameworkDescription`. So we have the same problems as in dotnet/BenchmarkDotNet#448.

**Comment by romansp** [comment]
I'd like to work on this.

I already looked into what BenchmarkDotNet guys did to detect correct version of .NET Core. Here is the specific commit: https://github.com/dotnet/BenchmarkDotNet/commit/857f2bf92458d2b6321a5a61eb0e6a61d1d10bdf. Do you think it will be a good approach to fix this issue?

I found several places where `RuntimeInformation.FrameworkDescription` is currently used:
https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/nunitlite/OutputWriters/NUnit2XmlOutputWriter.cs#L126
https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/nunitlite/TextUI.cs#L175
https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/framework/Api/FrameworkController.cs#L378

My guess is that all of these cases should be addressed in the PR. What do you think?

**Comment by jnm2** [comment]
@romansp Thank you, that is much appreciated!

Real quick, I have a question for you and @nunit/framework-team.

We've been showing "Framework version" and calling it "CLR version." To me that's confusing because the CLR is not a framework, it is a runtime, and specifically it is the Windows-only OS component.

.NET runtimes are virtual machines. Examples: CLR, CoreCLR, Mono, Blazor.
.NET frameworks are base class libraries. Examples: "The original" .NET Framework (netfx), .NET Core (corefx), Xamarin, UWP, Unity.

Do we intend to show the current framework version or the current runtime version, given that one does not always imply the other?

----

@romansp Back to the implementation:

https://github.com/dotnet/corefx/issues/23197 may yield something in the future, but for now, that approach is the best I know of. I'm thinking it would be good to keep it a statically-initialized property so that it is only calculated once.

Yes, all occurrences should be replaced with a call to the new implementing property.  Just spotted these which I think should be also replaced if possible:

https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/nunitlite/OutputWriters/NUnit2XmlOutputWriter.cs#L140-L141

 https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/nunitlite/TextUI.cs#L178

https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/framework/Api/FrameworkController.cs#L393

**Comment by ChrisMaddock** [comment]
@jnm2 - Why not both?! ¯\_(ツ)_/¯ Runtime is also user configurable, through the `--framework` option, so definitely should be shown somewhere. I think at the moment, you can maybe only see what you requested, not what actually run? I could be wrong here.

**Comment by jnm2** [comment]
@ChrisMaddock Sure, I'm happy with that! Determining the runtime version would require a little research and so is outside the scope of this particular issue but we can start another. In the meantime for this issue then, as @romansp adds the code to determine the actual framework version, can we change every occurrence of `"CLR Version"` to `"Framework version"`?

**Comment by ChrisMaddock** [comment]
Sounds good visually. For the xml - it could be possible that runners e.g. nunit-console are looking for the clr-version attribute to parse and display in the view. We should consider backwards-compat here - leaving a clr-version attribute in place may be preferable in the XML - even if it's not very 'correct'.

**Comment by jnm2** [comment]
> We should consider backwards-compat here

Oh, I was thinking of them all as display strings but you're absolutely right. XML should stay `<clr-version>` while display labels can be `Framework Version`.

@romansp, can you also change these two lines to say `Framework Version`?

https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/nunitlite/TextUI.cs#L175

https://github.com/nunit/nunit/blob/8e71ca642f2033fdd6cf2976c8f165824955dc74/src/NUnitFramework/nunitlite/TextUI.cs#L178

**Comment by romansp** [comment]
Thanks guys for all the input. Going to dedicate my time on the weekend to make a good PR. 

@jnm2, sure! Will change `Environment.Version` usages that you mentioned, thanks.

**Comment by romansp** [comment]
Sorry guys, unfortunately can't find any spare time to work on this issue anytime soon. I wouldn't mind if someone else wants to take it.

**Comment by mikkelbu** [comment]
@romansp It is completely okay. I happens for all of us. I've removed you from the issue. And thanks for making the issue more accessible 👍

**Comment by greentests** [comment]
Can I work on this as my first contribution?

**Comment by CharliePoole** [comment]
@greentests You're welcome to give it a shot! Since there has been various discussion here about alternative approaches, please tell us what you intend before you do it. 😉 

I've spent a little time on this as part of another issue. Here are a few hints.

1. Historically, the console runner displayed the framework under which __it__ was running, rather than the tests. In fact, with the console runner, there can be multiple assemblies running under different frameworks.

2. With NUnitLite, it's the same thing __except__ that the tests actually run under the same platform as NUnitLite itself.

3. I think it's probably a bad idea that NUnitLite currently uses the framework class `RunTimeframework` to figure out the platform. I think that nunitlite and the framework should be a bit more standoffish in this respect. 😄 

4. For the .NET Standard 1.6 build, you can probably use `AppContext` to get the framework name. This will give you a better .NET Core version. For .NET Standard 1.3, that's not available so you can only use `RuntimeInformation.TargetFrameworkDescription`, which will give you the 4.0 version. Of course, with research you may find something better but even if you only solve this for 1.6 it will be an advance.

**Comment by greentests** [comment]
Thanks for the useful hints @CharliePoole. I am researching a bit after going through the discussions and alternatives suggested about what could be the approaches. Meanwhile, I am also trying to figure out why I am not able to build nunit.sln even after installing VS2017. 
I would be posting the alternatives before weekend.

**Comment by jnm2** [comment]
@greentests We are about to address that, but for now, run `.\build.ps1 -t build` and the build in Visual Studio will start working. (It's not restoring NuGet packages correctly.)

**Comment by greentests** [comment]
Thanks @jnm2 it worked. I can build now.

**Comment by stevenaw** [comment]
I think this is no longer an issue. It looks like this is now driven by an assembly-level attribute:
https://github.com/nunit/nunit/blob/a5adf028c101283beecad0b49104468efe75b90f/src/NUnitFramework/nunitlite/TextUI.cs#L83-L87
https://github.com/nunit/nunit/blob/d83e7150edb41340d5d00e76e4fa918509fbfeb9/src/CommonAssemblyInfo.cs#L59-L63

And the output seems correct now:
![image](https://user-images.githubusercontent.com/439844/97822889-4774d800-1c85-11eb-923f-c236ae2f5e8e.png)

**Comment by jnm2** [comment]
Cool, thanks @stevenaw!

**Comment by ChrisMaddock** [comment]
Hi all - I think this issue is actually still outstanding. Funnily enough I upgraded the NUnit Engine .NET Standard 2.0 tests to run on the latest master of NUnitLite just yesterday!

Below is a screenshot of what the current master looks like running on .NET Core 2.1, from the nunit-console logs.

![image](https://user-images.githubusercontent.com/4837132/100019308-ba312900-2dd5-11eb-9eda-ae67d5e30187.png)

Looks like the wording has changed from what [@]jnm2 originally posted, but this still shows `CLR Version: 4.0.30319.42000`, which I think is the issue? Looks like that comes from this line:

https://github.com/nunit/nunit/blob/a5adf028c101283beecad0b49104468efe75b90f/src/NUnitFramework/nunitlite/TextUI.cs#L174

FWIW, this is how the Console currently shows this info - I guess we'd want NUnitLite to do the same? CLR version of course doesn't make sense for running on .NET Core. It is hopefully an easyfix now though, however. 🙂 

![image](https://user-images.githubusercontent.com/4837132/100019596-3af02500-2dd6-11eb-80e2-c3a63d42d3cf.png)

**Comment by stevenaw** [comment]
Oh sorry about that @ChrisMaddock ! Thanks for catching and pointing out the line that needs updating

**Comment by ChrisMaddock** [comment]
No worries! 😊

**Comment by sdelarosbil** [comment]
I looked at this and I think I need to ask a bit more info on what we want to do with this because this is more confusing than I realised.

It turns out that @ChrisMaddock 's outputs showed something a bit confusing that has nothing to do with nunit: Environement.Version only returns the actual runtime version (NOT the CLR version) starting with .net core 3.0. This is documented here: https://learn.microsoft.com/en-us/dotnet/api/system.environment.version?view=net-8.0#remarks

This is why his nunitlite running with .net core 2.1 showed a CLR version of 4.x, but running the console runner on .net core 3.1, we get the runtime version. You can actually have the console runner say 4.x if you run it on .net core 2.1 too meaning the version number issue isn't exclusive to nunit lite, but rather .net core itself.

Interestingly though, the console runner seems to understand this by doing the following:
```cs
#if NETFRAMEWORK
            OutWriter.WriteLabelLine(INDENT4 + "Runtime: ", ".NET Framework CLR v" + Environment.Version.ToString());
#else
``` But that would exclude .net core versions before 3.0 which outputs the CLR version.

So, bottom line is, I am not sure what we can do. The assembly attribute mentioned above concerns the version the runner was built on, but the runtime could not necessarily match. While we could do what was suggested to change the word "CLR" to "Runtime", it still means that for any .net core before 3.0, we will see 4.x.

**Comment by CharliePoole** [comment]
FWIW, the original purpose of all those `#if`s was to make it easy to see which build of nunitlite and the framework was in use. Prior to that, we would spend a lot of time on reported issues trying to determine the build in use. Of course, use of assembly attributes can _only_ tell you how the assembly was built, not the platform under which it is running.
